server/provider/nodeStatus.go

```diff
@@ -0,0 +1,70 @@
+package provider
+
+import "github.com/ElrondNetwork/rosetta/server/resources"
+
+// GetNodeStatus gets an aggregated node status (e.g. current block, oldest available block etc.)
+func (provider *networkProvider) GetNodeStatus() (*resources.AggregatedNodeStatus, error) {
+	if provider.isOffline {
+		return nil, errIsOffline
+	}
+
+	plainNodeStatus, err := provider.getPlainNodeStatus()
+	if err != nil {
+		return nil, err
+	}
+
+	latestNonce := getLatestNonceGivenHighestFinalNonce(plainNodeStatus.HighestFinalNonce)
+	latestBlockSummary, err := provider.getBlockSummaryByNonce(latestNonce)
+	if err != nil {
+		return nil, err
+	}
+
+	oldestNonceWithHistoricalState := provider.getOldestNonceWithHistoricalStateGivenNodeStatus(plainNodeStatus)
+	oldestBlockWithHistoricalState, err := provider.getBlockSummaryByNonce(oldestNonceWithHistoricalState)
+	if err != nil {
+		return nil, err
+	}
+
+	return &resources.AggregatedNodeStatus{
+		Synced:                         plainNodeStatus.IsSyncing == 0,
+		LatestBlock:                    latestBlockSummary,
+		OldestBlockWithHistoricalState: oldestBlockWithHistoricalState,
+	}, nil
+}
+
+func (provider *networkProvider) getPlainNodeStatus() (*resources.NodeStatus, error) {
+	if provider.isOffline {
+		return nil, errIsOffline
+	}
+
+	response := &resources.NodeStatusApiResponse{}
+	err := provider.getResource(urlPathGetNodeStatus, response)
+	if err != nil {
+		return nil, err
+	}
+
+	return &response.Data.Status, nil
+}
+
+func (provider *networkProvider) getLatestBlockNonce() (uint64, error) {
+	nodeStatus, err := provider.getPlainNodeStatus()
+	if err != nil {
+		return 0, err
+	}
+
+	// In the context of scheduled transactions, make sure the N+1 block is final, as well.
+	return getLatestNonceGivenHighestFinalNonce(nodeStatus.HighestFinalNonce), nil
+}
+
+func getLatestNonceGivenHighestFinalNonce(highestFinalNonce uint64) uint64 {
+	return highestFinalNonce - 1
```

Comment:
What is the reason we consider latest nonce to be one nonce below highestFinalNonce?
is there some limitation?

is it what we consider the node to still have trie data for?

Reply:
The N+1 block (handled in the context of scheduled miniblocks) has to be final as well. So that all blocks seen by rosetta (including the N+1 block) are already final.

Reply:
ok got it